### ISSUE TITLE
Why does it restart the UI whenever you save a new config?

### ISSUE THREAD

**Issue by Jonseed** [Zyin055/Config-Presets#25]
Saving new styles doesn't require restarting the UI, why does saving a new config restart the UI? Can it not add the requisite lines to the JSON file, and refresh its list without needing a full UI restart?

**Comment by Zyin055** [comment]
There's some trickery going on with the code that was causing issues, and the only way I could figure out how to resolve it was to restart the UI after creating a new config preset. It should be possible to do it without a restart like before A1111's update a few weeks ago, but I couldn't get it to work after the update.

**Comment by Jonseed** [comment]
Maybe you could look at the way styles are being saved, and copy that method?

**Comment by Zyin055** [comment]
It's not quite that simple. The styles button reads the values from predefined 2 textboxes (prompt/neg prompt), but config presets needs to accept a variable number of arguments because it works with both txt2img and img2img tabs. To do this we need to make use of a coding trick with *args and anonymous functions. Combined with Gradio being really stupid with how their dropdowns work, desyncs happen between server and client, and I couldn't find a way to fix that desync.

**Comment by medledan** [comment]
> It's not quite that simple. The styles button reads the values from predefined 2 textboxes (prompt/neg prompt), but config presets needs to accept a variable number of arguments because it works with both txt2img and img2img tabs. To do this we need to make use of a coding trick with *args and anonymous functions. Combined with Gradio being really stupid with how their dropdowns work, desyncs happen between server and client, and I couldn't find a way to fix that desync.

Can you just get it to save the config and let us know its not usable until we restart the UI? I'm usually in the middle of something and restarting the UI interrupts me so I rarely ever create any preconfigs.

**Comment by Symbiomatrix** [comment]
@medledan You need to comment out this function (line ~600) in config_presets (add `#` before each line):
```
save_button.click(  # need this to runa after save_config()
                             fn=None,
                             _js="config_preset_settings_restart_gradio()",  # restart Gradio
                             )
```
All new presets will show up after a reset, and existing preset changes will take effect immediately, which isn't half bad.

@Zyin055 I've been plagued by basically the same thing in regional prompter's presets - the menu item is added to the dropdown with gr.update, but it isn't recognised by the selection event. Are you aware of any fix in recent gradio versions / issue opened in the repo on the matter?

**Comment by Symbiomatrix** [comment]
So I think I've located the cause for this in https://github.com/gradio-app/gradio/issues/4210 ; the simple workaround is updating the dropdown with a persistent list object, same one its choices are inited with. You can check the PR / commit in regional prompter for details, my testing has shown it's working perfectly so far, no restart needed.

**Comment by Zyin055** [comment]
Fixed with PR #41
